cmd/zc_traverser_s3.go

```diff
@@ -0,0 +1,146 @@
+// Copyright © 2019 Microsoft <wastore@microsoft.com>
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package cmd
+
+import (
+	"context"
+	"fmt"
+	"net/url"
+	"strings"
+
+	"github.com/minio/minio-go"
+
+	"github.com/Azure/azure-storage-azcopy/common"
+)
+
+type s3Traverser struct {
+	rawURL    *url.URL // No pipeline needed for S3
+	ctx       context.Context
+	recursive bool
+
+	s3URLParts s3URLPartsExtension
+	s3Client   *minio.Client
+
+	// A generic function to notify that a new stored object has been enumerated
+	incrementEnumerationCounter func()
+}
+
+func (t *s3Traverser) traverse(processor objectProcessor, filters []objectFilter) (err error) {
+	// Check if resource is a single object.
+	if t.s3URLParts.IsObjectSyntactically() && !t.s3URLParts.IsDirectorySyntactically() && !t.s3URLParts.IsBucketSyntactically() {
+		objectPath := strings.Split(t.s3URLParts.ObjectKey, "/")
+		objectName := objectPath[len(objectPath)-1]
+
+		oi, err := t.s3Client.StatObject(t.s3URLParts.BucketName, t.s3URLParts.ObjectKey, minio.StatObjectOptions{})
+
+		// If we actually got object properties, process them.
+		// Otherwise, treat it as a directory.
+		// According to IsDirectorySyntactically, objects and folders can share names
+		if err == nil {
+			err = processIfPassedFilters(
+				filters,
+				newStoredObject(
+					objectName,
+					"", // We already know the exact path -- No need.
+					oi.LastModified,
+					oi.Size,
+					nil,
+					blobTypeNA,
+					""), // We already know the bucket name -- no need.
+				processor)
+
+			if err != nil {
+				return err
+			}
+
+			return nil
+		}
+	}
+
+	// Append a trailing slash if it is missing.
+	if !strings.HasSuffix(t.s3URLParts.ObjectKey, "/") && t.s3URLParts.ObjectKey != "" {
+		t.s3URLParts.ObjectKey += "/"
+	}
+
+	// Ignore *s in URLs and treat them as normal characters
```

Comment:
Worth adding a "why comment" here to say, "... because * can be a real URL character and, after discussion, we decided to treat * in URL as an ordinary character for that reason - i.e. not a wildcard".   Generally, if was significant or non-obvious enough that we had to discuss the reason why, then its worth including the reason why in a comment.